functionGenerateHeapSnapshot should expect an OOM exception in JSONParse

ast-hugger opened the pull request:
#### 6b0c696557568ae8aa82875fcac5fbd496510ab4
<pre>
functionGenerateHeapSnapshot should expect an OOM exception in JSONParse
<a href="https://rdar.apple.com/150692273">rdar://150692273</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294991">https://bugs.webkit.org/show_bug.cgi?id=294991</a>

Reviewed by Keith Miller.

The crash reported in the radar is an assertion failure that may happen in
`functionGenerateHeapSnapshot` when it calls `JSONParse` to parse a json string. The call
is followed by an assertion that no exception was thrown. However, in low memory
conditions parsing may legitimately throw an OutOfMemory exception. This happens in ASAN
builds when they run the test case reported in the radar. In non-ASAN builds OOM is not
thrown when running the same test case.

The patch changes the shapshot function to expect and propagate exceptions thrown in the
parse call.

* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/296739@main">https://commits.webkit.org/296739@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b429fb751d3ae99b849cdbde457dcc7c20fd17a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109318 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19405 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114524 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59574 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29655 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37565 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83082 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23606 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98458 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63534 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22995 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16601 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59150 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101825 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92974 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16643 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117637 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107876 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36359 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26910 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92093 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36732 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94720 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91904 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23424 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36830 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14580 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32182 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36255 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41740 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132143 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35936 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35809 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39267 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37631 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/04e3e3fb100e30229f8ece088f1d008cda825a93. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47205

hyjorc1 commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3179,7 +3179,7 @@ JSC_DEFINE_HOST_FUNCTION(functionGenerateHeapSnapshot, (JSGlobalObject* globalOb
 
     EncodedJSValue result = JSValue::encode(JSONParse(globalObject, jsonString));
-    scope.releaseAssertNoException();
+    RETURN_IF_EXCEPTION(scope, { });
```
Maybe we should use `RELEASE_AND_RETURN` here?

hyjorc1 replied:
Yeah, we should just do `RELEASE_AND_RETURN(scope, JSValue::encode(JSONParse(globalObject, jsonString)));` since `JSONParse` can throw exceptions. And we should also fix the other `JSONParse` uses in our code base.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/05a29d57124df55c2e44980a51f043139c94095c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47205

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6b429fb751d3ae99b849cdbde457dcc7c20fd17a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47205

ast-hugger committed:
functionGenerateHeapSnapshot should expect an OOM exception in JSONParse
rdar://150692273
https://bugs.webkit.org/show_bug.cgi?id=294991

Reviewed by Keith Miller.

The crash reported in the radar is an assertion failure that may happen in
`functionGenerateHeapSnapshot` when it calls `JSONParse` to parse a json string. The call
is followed by an assertion that no exception was thrown. However, in low memory
conditions parsing may legitimately throw an OutOfMemory exception. This happens in ASAN
builds when they run the test case reported in the radar. In non-ASAN builds OOM is not
thrown when running the same test case.

The patch changes the shapshot function to expect and propagate exceptions thrown in the
parse call.

* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/296739@main
